src/audit/self_hosted_runner.rs

```diff
@@ -106,21 +106,171 @@ impl WorkflowAudit for SelfHostedRunner {
+                        exp.as_bare() == path && expansion.contains("self-hosted")
+                    });
+
+                    if self_hosted {
+                        results.push(
+                            Self::finding()
+                                .confidence(Confidence::High)
+                                .severity(Severity::Unknown)
+                                .persona(Persona::Auditor)
+                                .add_location(
+                                    job.location()
+                                        .with_keys(&["strategy".into()])
+                                        .annotated("matrix declares self-hosted runner"),
+                                )
+                                .add_location(
+                                    job.location().with_keys(&["runs-on".into()]).annotated(
+                                        "expression may expand into a self-hosted runner",
+                                    ),
+                                )
+                                .build(workflow)?,
                         )
-                        .build(workflow)?,
-                ),
+                    }
+                }
             }
         }
 
         Ok(results)
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::audit::self_hosted_runner::SelfHostedRunner;
+    use crate::audit::WorkflowAudit;
+    use crate::models::Workflow;
+    use crate::state::{AuditState, Caches};
+    use camino::Utf8Path;
+    use indoc::indoc;
+    use std::fs;
+    use temp_dir::TempDir;
+    use uuid::Uuid;
+
+    fn parsed_workflow(contents: &str) -> Workflow {
+        let temp_dir = TempDir::new().expect("Cant create temp dir");
+        let file_name = Uuid::new_v4().to_string();
+        let target = temp_dir.path().join(format!("{}.yml", file_name));
+        fs::write(&target, contents).expect("Failed to write content to sut markdown file");
+        let utf8_path = Utf8Path::from_path(target.as_path()).unwrap();
+        Workflow::from_file(utf8_path).expect("invalid GitHub Actions workflow")
+    }
+
+    fn self_hosted_runner_audit() -> SelfHostedRunner {
+        let audit_state = AuditState {
+            no_online_audits: true,
+            gh_token: None,
+            caches: Caches::new(),
+        };
+
+        SelfHostedRunner::new(audit_state).expect("invalid audit state")
+    }
+
+    #[test]
+    fn self_hosted_on_labels() {
+        let workflow = indoc! {"
+            on:
+              push:
+
+            jobs:
+              whops:
+                runs-on: [self-hosted, linux, arm64]
+
+                steps:
+                  - run: echo \"hello from a self-hosted runner\"
+        "};
+
+        let auditable = parsed_workflow(workflow);
+
+        let sut = self_hosted_runner_audit();
+
+        let findings = sut.audit(&auditable).expect("cannot audit workflow");
+
+        assert_eq!(findings.len(), 1);
```

Comment:
Because we are running only one audit, I'm just asserting that we find something. Happy to improve here, although constructing a `Finding` instance is not that simple. We can assert determinations with ease, though

Reply:
Let's use snapshot tests for these! That'll be easier to maintain and doesn't require us to have inline workflows or construct `Finding` objects here 🙂 

(These could all be individual workflows under `test-data` (or `test-data/matrices` or similar), or they could be one big workflow. No preference from me.)

Reply:
I'll do that, I had the current test written I got `insta` stuff only after syncing 🙂

Reply:
Done [here](https://github.com/woodruffw/zizmor/pull/231/commits/e28d18fc1b7c873c2079be946a87bc6f2fd16eaf), I added a sub-folder `self-hosted`, but happy to add more test cases covering `template-injection` 🙂

Reply:
I went for those tests and caught a bug [here](https://github.com/woodruffw/zizmor/pull/231/commits/686b9665804331658b7315ffe8356238bbd8f841) 😄

Reply:
Nice!

Reply:
Yeah, subfolders are totally fine by me.